versioned/storage/store/src/main/java/org/projectnessie/versioned/storage/versionstore/CommitImpl.java

```diff
@@ -186,11 +199,29 @@ CommitResult<Commit> commit(
 
     fromCommitMeta(metadata, commit);
 
+    CommitObj newHead;
     try {
-      CommitObj newHead = commitLogic.doCommit(commit.build(), objectsToStore);
-
+      CreateCommit createCommit = commit.build();
+      newHead =
+          commitLogic.buildCommitObj(
+              createCommit,
+              c -> CONFLICT,
+              (k, v) -> {},
+              NO_VALUE_REPLACEMENT,
+              NO_VALUE_REPLACEMENT);
+
+      // If 'commitRetryState.storedContents' already contains the commit-ID, __we__ already
+      // successfully persisted that commit. This can happen, if the `Persist` implementation raised
+      // an `UnknownOperationResultException` in one of the following operations (reference bump for
+      // example).
+      boolean stored = commitLogic.storeCommit(newHead, objectsToStore);
+      if (stored) {
+        commitRetryState.commitPersisted = newHead.id();
+        commitRetryState.storedContents.addAll(toStore);
+      }
+      boolean commitPersisted = stored || newHead.id().equals(commitRetryState.commitPersisted);
```

Comment:
When can `newHead.id()` equal `commitRetryState.commitPersisted` if `stored == false`?

I suppose `newHead` is unique in every call to `buildCommitObj` if the microsecond clock ticks (which is mostly a given in our retry loop, I guess).

Reply:
It's not unique, think how commit-obj-IDs are generated.

Reply:
My mistake... I thought the microsecond commit time was part of the commit hash (but it is not) :facepalm: